app/scripts/lib/multichain-api/wallet-requestPermissions.js

```diff
@@ -0,0 +1,179 @@
+import { parseCaipAccountId } from '@metamask/utils';
+import {
+  CaveatTypes,
+  RestrictedMethods,
+} from '../../../../shared/constants/permissions';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from './caip25permissions';
+import { mergeScopes } from './scope';
+
+export const requestPermissionsHandler = {
+  methodNames: [MethodNames.requestPermissions],
+  implementation: requestPermissionsImplementation,
+  hookNames: {
+    requestPermissionsForOrigin: true,
+    getPermissionsForOrigin: true,
+    getNetworkConfigurationByNetworkClientId: true,
+    updateCaveat: true,
+    grantPermissions: true,
+  },
+};
+
+/**
+ * Request Permissions implementation to be used in JsonRpcEngine middleware.
+ *
+ * @param req - The JsonRpcEngine request
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options - Method hooks passed to the method implementation
+ * @param options.requestPermissionsForOrigin - The specific method hook needed for this method implementation
+ * @param options.getPermissionsForOrigin
+ * @param options.getNetworkConfigurationByNetworkClientId
+ * @param options.updateCaveat
+ * @param options.grantPermissions
+ * @returns A promise that resolves to nothing
+ */
+async function requestPermissionsImplementation(
+  req,
+  res,
+  _next,
+  end,
+  {
+    requestPermissionsForOrigin,
+    getPermissionsForOrigin,
+    getNetworkConfigurationByNetworkClientId,
+    updateCaveat,
+    grantPermissions,
+  },
+) {
+  const { origin, params } = req;
+
+  if (!Array.isArray(params) || !isPlainObject(params[0])) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  const [requestedPermissions] = params;
+  delete requestedPermissions[Caip25EndowmentPermissionName];
+
+  const [_grantedPermissions] = await requestPermissionsForOrigin(
+    requestedPermissions,
+  );
+
+  // caveat values are frozen and must be cloned before modified
+  const grantedPermissions = { ..._grantedPermissions };
+
+  const ethAccountsPermission =
+    grantedPermissions[RestrictedMethods.eth_accounts];
+
+  if (process.env.BARAD_DUR && ethAccountsPermission) {
+    // TODO: Use permittedChains permission returned from requestPermissionsForOrigin() when available
+    const { chainId } = getNetworkConfigurationByNetworkClientId(
+      req.networkClientId,
+    );
+
+    const scopeString = `eip155:${parseInt(chainId, 16)}`;
+
+    const ethAccounts = ethAccountsPermission.caveats[0].value;
+
+    const caipAccounts = ethAccounts.map(
+      (account) => `${scopeString}:${account}`,
+    );
+
+    const permissions = getPermissionsForOrigin(origin);
+    const caip25endowment = permissions[Caip25EndowmentPermissionName];
+    const caip25caveat = caip25endowment?.caveats.find(
+      ({ type }) => type === Caip25CaveatType,
+    );
+    if (caip25caveat) {
```

Comment:
?
```suggestion
    const caip25Endowment = permissions[Caip25EndowmentPermissionName];
    const caip25Caveat = caip25Endowment?.caveats.find(
      ({ type }) => type === Caip25CaveatType,
    );
    if (caip25Caveat) {
```

Reply:
i messed up this casing in a lot of places in our feature branch already. Let's do this in another PR. I'll add a note to the slack board